pennylane/ops/op_math/controlled_class.py

```diff
@@ -0,0 +1,477 @@
+
+        canonical_matrix = qmlmath.block_diag([left_pad, base_matrix, right_pad])
+
+        if wire_order is None or self.wires == Wires(wire_order):
+            return canonical_matrix
+
+        active_wires = self.control_wires + self.target_wires
+        return operation.expand_matrix(canonical_matrix, wires=active_wires, wire_order=wire_order)
+
+    # pylint: disable=arguments-differ
+    def sparse_matrix(self, wire_order=None, format="csr"):
+        if wire_order is not None:
+            raise NotImplementedError("wire_order argument is not yet implemented.")
+
+        try:
+            target_mat = self.base.sparse_matrix()
+        except operation.SparseMatrixUndefinedError:
+            try:
+                target_mat = sparse.lil_matrix(self.base.matrix())
+            except operation.MatrixUndefinedError as e:
+                raise operation.SparseMatrixUndefinedError from e
+
+        num_target_states = 2 ** len(self.target_wires)
+        num_control_states = 2 ** len(self.control_wires)
+        total_states = num_target_states * num_control_states
+
+        start_ind = self._control_int * num_target_states
+        end_ind = start_ind + num_target_states
+
+        m = sparse.eye(total_states, format="lil", dtype=target_mat.dtype)
+
+        m[start_ind:end_ind, start_ind:end_ind] = target_mat
+
+        return m.asformat(format=format)
+
+    def eigvals(self):
+        base_eigvals = self.base.eigvals()
+        num_target_wires = len(self.target_wires)
+        num_control_wires = len(self.control_wires)
+
+        total = 2 ** (num_target_wires + num_control_wires)
+        ones = np.ones(total - len(base_eigvals))
+
+        return qmlmath.concatenate([ones, base_eigvals])
+
+    def diagonalizing_gates(self):
+        return self.base.diagonalizing_gates()
+
+    def decomposition(self):
+        if not all(self.control_values):
+            d = [
+                qml.PauliX(w) for w, val in zip(self.control_wires, self.control_values) if not val
+            ]
+            d += [Controlled(self.base, self.control_wires, work_wires=self.work_wires)]
+            d += [
+                qml.PauliX(w) for w, val in zip(self.control_wires, self.control_values) if not val
+            ]
+
+            return d
+        # More to come.  This will be an extensive PR in and of itself.
+        return super().decomposition()
+
+    def generator(self):
+        sub_gen = self.base.generator()
+        proj_tensor = operation.Tensor(*(qml.Projector([1], wires=w) for w in self.control_wires))
+        return 1.0 * proj_tensor @ sub_gen
+
+    def adjoint(self):
+        return Controlled(
+            self.base.adjoint(), self.control_wires, self.control_values, self.work_wires
+        )
+
+    def pow(self, z):
+        base_pow = self.base.pow(z)
+        return [
+            Controlled(op, self.control_wires, self.control_values, self.work_wires)
+            for op in base_pow
+        ]
```

Comment:
`base.pow(z)` returns an iterable? I thought it would just be the Pow wrapper on the base: 

Something like `Pow(Controlled(base)) --> Controlled(Pow(base))`

Reply:
The `pow` method is essentially the decomposition for the `Pow` class containing an operator of that type.  It's a shortcut to an easier way of computing the power, so it should not return something with the `Pow` class.

When we get around to writing up a "simplification" procedure, then we can start thinking about things like `Pow(Controlled(base)) --> Controlled(Pow(base))`, but I don't think we should do so here.